tests/Unit/Evolution/Systems/ScalarWave/Test_Characteristics.cpp

```diff
@@ -0,0 +1,388 @@
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& u_zero,
+    const Scalar<DataVector>& u_plus, const Scalar<DataVector>& u_minus,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& unit_normal_one_form) {
+  return get<Tag>(
+      ScalarWave::EvolvedFieldsFromCharacteristicFieldsCompute<Dim>::function(
+          gamma_2, u_psi, u_zero, u_plus, u_minus, unit_normal_one_form));
+}
+
+template <size_t Dim>
+void test_evolved_from_characteristic_fields() noexcept {
+  const DataVector used_for_size(20);
+  // Psi
+  pypp::check_with_random_values<1>(evol_field_with_tag<ScalarWave::Psi, Dim>,
+                                    "Characteristics", "evol_field_psi",
+                                    {{{-100., 100.}}}, used_for_size);
+  // Pi
+  pypp::check_with_random_values<1>(evol_field_with_tag<ScalarWave::Pi, Dim>,
+                                    "Characteristics", "evol_field_pi",
+                                    {{{-100., 100.}}}, used_for_size);
+  // Phi
+  pypp::check_with_random_values<1>(
+      evol_field_with_tag<ScalarWave::Phi<Dim>, Dim>, "Characteristics",
+      "evol_field_phi", {{{-100., 100.}}}, used_for_size);
+}
+
+// Test return-by-reference fundamental fields by comparing to analytic solution
+template <size_t Dim, typename Solution>
+void test_evolved_from_characteristic_fields_analytic(
+    const Solution& solution, const size_t grid_size_each_dimension,
+    const std::array<double, 3>& lower_bound,
+    const std::array<double, 3>& upper_bound) noexcept {
+  // Set up grid
+  Mesh<Dim> mesh{grid_size_each_dimension, Spectral::Basis::Legendre,
+                 Spectral::Quadrature::GaussLobatto};
+
+  using Affine = domain::CoordinateMaps::Affine;
+  using Affine3D =
+      domain::CoordinateMaps::ProductOf3Maps<Affine, Affine, Affine>;
+  const auto coord_map =
+      domain::make_coordinate_map<Frame::Logical, Frame::Inertial>(Affine3D{
+          Affine{-1., 1., lower_bound[0], upper_bound[0]},
+          Affine{-1., 1., lower_bound[1], upper_bound[1]},
+          Affine{-1., 1., lower_bound[2], upper_bound[2]},
+      });
+
+  // Set up coordinates
+  const auto x_logical = logical_coordinates(mesh);
+  const auto x = coord_map(x_logical);
+  // Arbitrary time for time-dependent solution.
+  const double t = 0.;
+
+  // Evaluate analytic solution
+  const auto vars = solution.variables(
+      x, t,
+      tmpl::list<ScalarWave::Pi, ScalarWave::Phi<Dim>, ScalarWave::Psi>{});
+  // Get ingredients
+  const size_t n_pts = x.begin()->size();
+  const auto gamma_2 = make_with_value<Scalar<DataVector>>(x, 0.1);
+  const auto& psi = get<ScalarWave::Psi>(vars);
+  const auto& phi = get<ScalarWave::Phi<Dim>>(vars);
+  const auto& pi = get<ScalarWave::Pi>(vars);
+  // Outward 3-normal to the surface on which characteristic fields are needed
+  auto unit_normal_one_form =
+      make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(x, 1.);
```

Comment:
Again, could you directly construct this with the right value?

Reply:
done